src/main/ApplicationUtils.cpp

```diff
@@ -477,6 +477,278 @@ processArchivalMetrics(
+        {
+            return 0;
+        }
+
+        auto const& val = le.data.contractData().val;
+        if (val.type() == SCV_MAP && val.map() && val.map()->size() != 0)
+        {
+            auto const& amountEntry = val.map()->at(0);
+            if (amountEntry.key == assetContractInfo.mAmountSymbol)
+            {
+                if (amountEntry.val.type() == SCV_I128)
+                {
+                    auto lo = amountEntry.val.i128().lo;
+                    auto hi = amountEntry.val.i128().hi;
+                    if (lo > INT64_MAX || hi > 0)
+                    {
+                        throw std::runtime_error(
+                            "Asset contractData balance amount out of range");
+                    }
+                    return static_cast<int64_t>(lo);
+                }
+            }
+        }
+        return 0;
+    }
+    case CONTRACT_CODE:
+        break;
+    case CONFIG_SETTING:
+        break;
+    case TTL:
+        break;
+    }
+    return 0;
+}
+
+void
+getHotArchiveListBalanceForAsset(Application& app,
+                                 HistoryArchiveState const& has,
+                                 Asset const& asset,
+                                 AssetContractInfo const& assetContractInfo,
+                                 int64_t& runningBalance)
+{
+    // hot archive buckets
+    std::vector<Hash> hotHashes;
+    for (uint32_t i = 0; i < has.hotArchiveBuckets.size(); ++i)
+    {
+        HistoryStateBucket<HotArchiveBucket> const& hsb =
+            has.hotArchiveBuckets.at(i);
+        hotHashes.emplace_back(hexToBin256(hsb.curr));
+        hotHashes.emplace_back(hexToBin256(hsb.snap));
+    }
+
+    auto& bm = app.getBucketManager();
+
+    std::unordered_set<LedgerKey> seenKeys;
+    for (auto const& hash : hotHashes)
+    {
+        if (isZero(hash))
+        {
+            continue;
+        }
+        auto b = bm.getBucketByHash<HotArchiveBucket>(hash);
+        if (!b)
+        {
+            throw std::runtime_error(std::string("missing bucket: ") +
+                                     binToHex(hash));
+        }
+
+        for (HotArchiveBucketInputIterator in(b); in; ++in)
+        {
+            auto const& be = *in;
+            if (be.type() != HOT_ARCHIVE_ARCHIVED)
+            {
+                if (be.type() == HOT_ARCHIVE_LIVE)
+                {
+                    seenKeys.emplace(be.key());
+                }
+                continue;
+            }
+            if (seenKeys.emplace(LedgerEntryKey(be.archivedEntry())).second ==
+                false)
+            {
+                // duplicate key, skip
+                continue;
+            }
+
+            auto balance =
+                getAssetBalance(be.archivedEntry(), asset, assetContractInfo);
+            if (!addBalance(runningBalance, balance))
+            {
+                throw std::runtime_error(
+                    "Total asset balance overflowed int64_t");
```

Comment:
Btw the overflows can happen here for non-XLM assets, can't they? So this is not quite generic, though probably serviceable for most of the real assets.

Reply:
Yeah I mention that this command only works for assets with total supply under `INT64_MAX`. This limitation can be removed with a little more work, but that can be done later if the need arises.